packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -945,3 +976,429 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+///
+///  * [OverlayEntry], the imperative API for inserting a widget to an [Overlay].
+class OverlayPortal extends RenderObjectWidget {
+
+  /// Creates a widget that renders the given `overlayChild` on the [Overlay]
+  /// specified by `overlayChild`.
+  const OverlayPortal({
+    super.key,
+    required this.overlayInfo,
+    required this.overlayChild,
+    required this.child,
+  }) : assert(overlayChild == null || overlayInfo != null);
+
+  /// A widget below this widget in the tree, but renders on the [Overlay]
+  /// given by [overlayInfo].
+  ///
+  /// The [overlayChild] widget, if not null, is placed below this widget in
+  /// the widget tree, allowing it to depend on [InheritedWidget] above it, and
+  /// be notified when the [InheritedWidget] changes.
+  ///
+  /// Unlike [child], [overlayChild] can visually extend outside the bounds
+  /// of this widget without being clipped, and receive hit-test events outside
+  /// of this widget's bounds, as long as it does not extend outside of the
+  /// [Overlay] it renders on.
+  final Widget? overlayChild;
+
+  /// A widget below this widget in the tree.
+  final Widget? child;
+
+  /// A location in a particular [Overlay].
+  ///
+  /// An [OverlayInfo] on the enclosing [Overlay] widget can be aquired using
+  /// [OverlayInfo.of].
+  ///
+  /// This parameter must not be null when [overlayChild] is not null.
+  final OverlayInfo? overlayInfo;
+
+  @override
+  RenderObjectElement createElement() => _OverlayPortalElement(this);
+
+  @override
+  RenderObject createRenderObject(BuildContext context) => _RenderLayoutSurrogateProxyBox();
+}
+
+class _OverlayPortalElement extends RenderObjectElement {
+  _OverlayPortalElement(OverlayPortal super.widget);
+
+  @override
+  OverlayPortal get widget => super.widget as OverlayPortal;
+
+  @override
+  _RenderLayoutSurrogateProxyBox get renderObject => super.renderObject as _RenderLayoutSurrogateProxyBox;
+
+  Element? _overlayChild;
+  Element? _child;
+
+  @override
+  void mount(Element? parent, Object? newSlot) {
+    super.mount(parent, newSlot);
+    _child = updateChild(_child, widget.child, null);
+    final Widget? overlayChild = widget.overlayChild;
+    _overlayChild = updateChild(
+      _overlayChild,
+      overlayChild == null ? null : _DeferredLayout(layoutSurrogate: renderObject, child: overlayChild),
+      widget.overlayInfo,
+    );
+  }
+
+  @override
+  void update(OverlayPortal newWidget) {
+    super.update(newWidget);
+
+    _child = updateChild(_child, newWidget.child, null);
+    final Widget? remoteChild = newWidget.overlayChild;
+    final Widget? wrappedRemoteChild = remoteChild == null
+      ? null
+      : _DeferredLayout(layoutSurrogate: renderObject, child: remoteChild);
+    _overlayChild = updateChild(_overlayChild, wrappedRemoteChild, newWidget.overlayInfo);
+  }
+
+  @override
+  void forgetChild(Element child) {
+    assert(child == _child);
+    // The _overlayChild Element does not have a key.
```

Comment:
this comment should explain why (because we create it ourselves in mount/update, without a key) and why it matters (because that's what forgetChild is about)